### PR TITLE
fix(optimizer): eliminate eliminate_unions

### PR DESCRIPTION
**Pull request by barakalon**
whew. This surfaced a few nasty bugs.

### CODE REVIEW

File: sqlglot/optimizer/merge_subqueries.py [hunk 1]
```diff
@@ -174,6 +174,26 @@ def _outer_select_joins_on_inner_select_join():
...
+    def _is_recursive():
+        # Recursive CTEs look like this:
+        #     WITH RECURSIVE cte AS (
+        #       SELECT * FROM ...
+        #       UNION ALL
+        #       SELECT *
+        #       FROM cte
+        #     )
+        # In this case, the inner scope is SELECT * FROM ..., NOT the entire CTE.
+        # So we traverse up to the inner scope's parent.
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py:186]: @barakalon what do you mean here? not sure I get it

File: tests/test_executor.py [hunk 2]
```diff
@@ -289,11 +289,6 @@ def test_set_operations(self):
                 ["a"],
                 [(1,), (2,), (3,)],
             ),
-            (
-                "SELECT 1 AS a UNION SELECT 2 AS a UNION SELECT 3 AS a",
-                ["a"],
-                [(1,), (2,), (3,)],
-            ),
```
Comment (georgesittas) [inline tests/test_executor.py:296]: Why was this removed?

**Review by georgesittas — COMMENTED** [review]
Awesome that you eliminated this 😄 🚀 

One question so that I can understand the recursive CTE change a bit better. Also I haven't dug deep into the TPC-DS test changes yet.

**Comment by georgesittas** [comment]
@barakalon it seems like this PR fixes another bug I found earlier..

Main:

```python
>>> execute("with t1 as (select 1 as x union all select 2 as x), t2 as (select 1 as x) select x from t1 union all select x from t2 limit 1")
Traceback (most recent call last):
  File "sqlglot/executor/python.py", line 39, in execute
    contexts[node] = self.scan(node, context)
  File "sqlglot/executor/python.py", line 105, in scan
    context, table_iter = self.scan_table(step)
  File "sqlglot/executor/python.py", line 133, in scan_table
    context = self.context({step.source.alias_or_name: table})
  File "sqlglot/executor/python.py", line 81, in context
    return Context(tables, env=self.env)
  File "sqlglot/executor/context.py", line 30, in __init__
    self.range_readers = {name: table.range_reader for name, table in self.tables.items()}
  File "sqlglot/executor/context.py", line 30, in <dictcomp>
    self.range_readers = {name: table.range_reader for name, table in self.tables.items()}
AttributeError: 'NoneType' object has no attribute 'range_reader'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sqlglot/executor/__init__.py", line 88, in execute
    result = PythonExecutor(tables=tables_).execute(plan)
  File "sqlglot/executor/python.py", line 61, in execute
    raise ExecuteError(f"Step '{node.id}' failed: {e}") from e
sqlglot.errors.ExecuteError: Step 'Scan: t2 (4380431776)' failed: 'NoneType' object has no attribute 'range_reader'
```

This branch:

```python
>>> execute("with t1 as (select 1 as x union all select 2 as x), t2 as (select 1 as x) select x from t1 union all select x from t2 limit 1")
x
1
2
1
```

Let's add this test as well? The reason it was previously failing was due to the optimizer producing incorrect SQL:

```python
>>> from sqlglot.optimizer import optimize
>>> print(optimize("with t1 as (select 1 as x union all select 2 as x), t2 as (select 1 as x) select x from t1 union all select x from t2 limit 1").sql(pretty=True))
WITH "cte" AS (
  SELECT
    2 AS "x"
), "t1" AS (
  SELECT
    "t2"."x" AS "x"
  FROM "t2" AS "t2"
  UNION ALL
  SELECT
    "cte"."x" AS "x"
  FROM "cte" AS "cte"
), "t2" AS (
  SELECT
    1 AS "x"
)
SELECT
  "t1"."x" AS "x"
FROM "t1"
UNION ALL
SELECT
  "t2"."x" AS "x"
FROM "t2"
LIMIT 1
```

**Comment by georgesittas** [comment]
> EDIT: actually, execute's result in this branch seems iffy as well, the LIMIT 1 should apply to the UNION and so I'd expect a single row to be returned..

https://github.com/tobymao/sqlglot/pull/2665

File: sqlglot/optimizer/merge_subqueries.py (hunk 1 above)
Comment (barakalon) [reply sqlglot/optimizer/merge_subqueries.py:186]: will update comment

File: tests/test_executor.py (hunk 2 above)
Comment (barakalon) [reply tests/test_executor.py:296]: Its a duplicate of the test directly above

Comment (georgesittas) [reply tests/test_executor.py:296]: oh missed that, nice catch 👍

**Comment by barakalon** [comment]
Righto - rebased

**Review by tobymao — APPROVED** [review]

File: sqlglot/optimizer/merge_subqueries.py [hunk 3]
```diff
@@ -174,6 +174,22 @@ def _outer_select_joins_on_inner_select_join():
...
+        #       SELECT * FROM x  <-- inner scope
+        #       UNION ALL
+        #       SELECT * FROM cte  <-- outer scope
+        #     )
+        cte = inner_scope.expression.parent
+        node = outer_scope.expression.parent
+
+        while node:
+            if node is cte:
+                return True
```
Comment (georgesittas) [inline sqlglot/optimizer/merge_subqueries.py:189]: Shouldn't we check that the `recursive` flag is set as well?

Comment (barakalon) [reply sqlglot/optimizer/merge_subqueries.py:189]: `recursive` is set on `With`, not `CTE`. We _can_, but I don't know if its necessary

Comment (georgesittas) [reply sqlglot/optimizer/merge_subqueries.py:189]: I guess I was a little confused about this loop/check being used as a proxy for detecting recursion because of the `_is_recursive` name. Is it sufficient? You can also have a CTE which is non-recursive and contains a `UNION` inside of it.

Comment (barakalon) [reply sqlglot/optimizer/merge_subqueries.py:189]: IIUC, a recursive WITH statement can include non-recursive ctes _and_ recursive ctes, so you might still want to merge.
If the outer scope is a child of the inner scope, you know you're recursive. That should be true for all scopes.

Comment (georgesittas) [reply sqlglot/optimizer/merge_subqueries.py:189]: I see, makes sense

**Review by georgesittas — APPROVED** [review]
LGTM

@barakalon do we cover the recursive cte condition in a test?

**Comment by barakalon** [comment]
Yep, theres an existing test
